File: Python Solutions/Easy/LongestCommonPrefix.py
# Author: Dorian Knight
# Date: December 16th 2024
# Description: Find the longest common prefix in a list of strings

class Solution:
    def longestCommonPrefix(self, strs: list[str]) -> str:
        # Find the shortest word in the list
        min_len = 201 # Maximum length according to the leetcode question specs
        for object in strs:
            if len(object) < min_len:
                min_len = len(object)

        # Iterate over each string in the string list and compare each character across all strings in order
        # strs[0][0] == strs[1][0] == strs[2][0]
        max_prefix_found = False
        max_prefix = ""
        for i in range(min_len):
            if len(strs) == 1:
                # Only one string in the input list
                max_prefix = strs[0]
                return max_prefix

            for j in range(len(strs)-1):
                charCurr = strs[j][i]
                charNext = strs[j+1][i]

                if charCurr != charNext:
                    max_prefix_found = True
                    return max_prefix

            if max_prefix_found == False:
                max_prefix += charCurr
        return max_prefix

def main():
    LCP = Solution()
    input = ["flow", "flower", "flight"]
    input = ["a"]
    answer = LCP.longestCommonPrefix(input)
    print(f"Input: {input} \nOutput: {answer}")

if __name__ == "__main__":
    main()

File: C# Solutions/Easy/LongestCommonPrefix.cs
// Author: Dorian Knight
// Date: December 24th 2024
// Description: Longest common prefix C# solution
// Rank: Easy

public class Solution {
    public string LongestCommonPrefix(string[] strs) {
        // Iterate through each string and see if the first "i" characters match.
        // If yes, keep going, if no, return the greatest common prefix
        string greatest_common_prefix = "";
        int iterator = 0;
        int min_length = 201;

        // Find the shortest word in the list
        foreach(string word in strs) {
            if (word.Length < min_length) {
                min_length = word.Length;
            }
        }

        // Iterate over every index across each word until you get to the last index in the shortest word
        bool compare_successful = true;
        while (iterator < min_length && compare_successful) {
            // Get the character at index "i" to compare against other characters in other strings at index "i"
            // If the comparison is successful then add that character to the prefix string
            char candidate_prefix_char = strs[0][iterator];

            foreach(string word in strs){
                char compare_against = word[iterator];
                if (candidate_prefix_char != compare_against) {
                    compare_successful = false;
                    break;
                }
            }

            if (compare_successful) {
                // String are immutable in C# :(
                greatest_common_prefix = greatest_common_prefix + candidate_prefix_char;
            }

            iterator++;
        }

        return greatest_common_prefix;
    }
}